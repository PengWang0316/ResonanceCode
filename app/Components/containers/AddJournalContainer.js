import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import JournalForm from '../JournalForm';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';
import { checkAuthentication } from '../../actions/UserActions';
import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import { updateJournal, createJournal, deleteJournal, deleteUnattachedJournal } from '../../actions/JournalActions';
import { clearReadings } from '../../actions/ReadingActions';

/**
 * Add Journal Container class.
 */
class AddJournalContainer extends Component {
  /** Checking authentication if user object cannot be found in props.
   * @returns {null} No return.
   */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    this.isSubmited = false; // Use this as an indicator for finishing submit.
  }

  /** After updating, creating and deleting, forward to reading page.
    * @param {object} nextProps is the object that contains new props.
    * @returns {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if (this.isSubmited && !this.props.journal && this.props.isLoading && !nextProps.isLoading) {
      this.props.clearReadings();
      this.props.history.push('/reading');
    }
  }

  /*
    ** updateObject = {
        isUpdate: bool,
        journalId: ,
        journalDate: ,
        userId: ,
        readings: ,
        contents: ,
        oldReadingIds:
      }
    */

  /** Handling the submit event.
   * @param {object} updateObject is the journal object will be updated or created.
   * @returns {null} No return.
   */
  handleSubmitCallback = updateObject => {
    // this.props.isLoading(true);
    const journal = Object.assign({
      _id: updateObject.journalId,
      date: new Date(updateObject.journalDate),
      // user_id: updateObject.userId,
      readings: updateObject.readings,
      shareList: updateObject.shareList
    }, updateObject.contents);
    // console.log("submit:", updateObject);
    if (updateObject.isUpdate) {
      // console.log("oldContentKeys:", this.oldContentKeys);
      // console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      // const deleteContents = [];
      const deleteReadingIds = [];
      // Object.prototype.hasOwnProperty.call(updateObject.readings) this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      // if this.oldReadingIds array is null, it is a unattached journal
      /* if (updateObject.oldReadingIds) updateObject.oldReadingIds.map((element) => { if (!updateObject.readings.hasOwnProperty(element)) deleteReadingIds.push(element); }); */
      if (updateObject.oldReadingIds) updateObject.oldReadingIds
        .forEach((element) => {
          if (!Object.prototype.hasOwnProperty.call(updateObject.readings, element))
            deleteReadingIds.push(element);
        });

      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      journal.isUnattachedJournal = !updateObject.oldReadingIds;
      // console.log("delete contents: ",deleteContents);
      // console.log("delete reading ids : ",deleteReadingIds);
      // updateJournal(journal, !updateObject.oldReadingIds).then((result) => {
      //   this.removeLoadingAndForward();
      // });
      this.props.updateJournal(journal);
    } else this.props.createJournal(journal);
    // this.removeLoadingAndForward();
    // createJournal(journal).then((result) => {
    //   this.removeLoadingAndForward();
    // });
    this.isSubmited = true;
  }

  /** Handling the delete event.
   * @param {string} journalId is the id for the journal.
   * @param {string} readingIds is a reading's id array that contains all reading connects to this journal.
   * @param {boolean} isUnattachedJournal is a indicator of whether is a unattached journal.
   * @returns {null} No return.
   */
  handleDeleteCallback = (journalId, readingIds, isUnattachedJournal) => {
    // console.log("Delete journal!");
    // event.preventDefault();
    // this.props.isLoading(true);
    this.props.clearJournalState();
    if (isUnattachedJournal) {
      this.props.deleteUnattachedJournal(journalId);
      this.props.history.push('/reading');
    } else this.props.deleteJournal({ journalId, readingIds });
    // deleteJournal(journalId, readingIds, userId).then(reault => {
    //   this.removeLoadingAndForward();
    // });
  }

  /*
  removeLoadingAndForward() {
    // this.props.isLoading(false);
    this.props.clearJournalState();
    this.props.history.push('/reading');
  }
*/
  /** Rendering the component.
   * @returns {jsx} Return the jsx that represents the component.
   */
  render() {
    // console.log("container: ", this.props.journal);
    return (
      <UnauthenticatedUserCheck>
        <div>
          <LoadingAnimation />
          <JournalForm
            journalData={this.props.journal}
            isWriting={this.props.isLoading}
            history={this.props.history}
            handleSubmit={this.handleSubmitCallback}
            handleDelete={this.handleDeleteCallback}
          />
        </div>
      </UnauthenticatedUserCheck>
    );
  }
}

/*
AddJournalContainer.propTypes = {
  journalData: PropTypes.object,
  history: PropTypes.object
}; */

const mapStateToProps = state => ({
  isLoading: state.isLoading,
  journal: state.journal,
  user: state.user
});

const mapDispatchToProps = dispatch => ({
  // isLoading: loadingState => dispatch(isLoading(loadingState)),
  checkAuthentication: _ => dispatch(checkAuthentication()),
  createJournal: journal => dispatch(createJournal(journal)),
  updateJournal: journal => dispatch(updateJournal(journal)),
  deleteJournal: params => dispatch(deleteJournal(params)),
  deleteUnattachedJournal: journalId => dispatch(deleteUnattachedJournal(journalId)),
  clearReadings: _ => dispatch(clearReadings())
});
export default connect(mapStateToProps, mapDispatchToProps)(AddJournalContainer);
