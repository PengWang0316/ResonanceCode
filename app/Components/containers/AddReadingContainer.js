import React, { Component } from 'react';
import { connect } from 'react-redux';
import jQuery from 'jquery';
import { JWT_MESSAGE } from '../../config';
import AddReadingForm from '../AddReadingForm';
// import Loading from '../Loading';
// import LoginApi from '../../apis/LoginApi';
import { getImageArray } from '../../apis/Util';
// import { createReading } from '../../apis/DatabaseApi';
import { checkAuthentication } from '../../actions/UserActions';

import { clickCoin, clearAddReadingTempState, createReading } from '../../actions/ReadingActions';
// import isLoading from '../../actions/LoadingActions';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';

/**
 * Add reading container
 * @returns {null} No return
 */
class AddReadingContainer extends Component {
  /**
  * Eliminating empty string from changeLines.
  * @param {string} changeLines is current change lines' string.
  * @returns {string} Return a new string without empty string.
  */
  static eliminateEmptyString(changeLines) {
    return changeLines.split(' ').filter((element) => element.length > 0).join(',');
  }

  /**
   * Handle subcomponent's cancel callback.
   * @returns {null} No return.
   */
  static handleCancelCallback = () =>
    $('#addReadingModal').modal('toggle'); // This method will use the $ that comes from index.html page.

  /**
   * Checking authentication
   * @returns {null} No return
   */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    this.props.clearAddReadingTempState();
  }

  /**
   * Clearing addReadingTempState, unlocking submit buttons, and turn off the addReadingForm (modal).
   * @param {nextProps} nextProps is an object that contains new props.
   * @returns {null} No return
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.readings.length !== 0 &&
      this.props.readings.length < nextProps.readings.length) {
      AddReadingContainer.handleCancelCallback();
      this.props.clearAddReadingTempState();
      jQuery('button[type=submit]').attr('disabled', '');
    }
  }

  /**
   * Handle subcomponent's coin click callback.
   * @param {int} lineNumber is the number of line.
   * @param {int} coinsPoint is the point that moment this line has based on the calculation.
   * @returns {null} No return.
   */
  handleCoinClickCallback = (lineNumber, coinsPoint) => {
    // console.log("LineNumber:",lineNumber);
    // console.log("coin point:",coinsPoint);
    const changeLinesArr = ['1st', '2nd', '3rd', '4th', '5th', '6th'];
    const regExpArr = [/1st/, /2nd/, /3rd/, /4th/, /5th/, /6th/];
    // change the state based on result
    const addReadingTempState = Object.assign({}, this.props.addReadingTempState); /* Making a copy from previous state */
    // console.log("Previous state: ", addReadingTempState);
    let side1;
    let middle1;
    let side2;
    let middle2;
    const classSide = 'img-line-side-big';
    const classMiddle = 'img-line-middle-big';
    const classMiddleB = 'img-line-middle-blank-big';
    // const classRed = 'background-red-big';
    const classRedSide = 'img-line-side-red-big';
    const classRedMiddle = 'img-line-middle-red-big';
    // console.log("b:",this.state.changeLines);
    // console.log("r:",this.changeLinesArr[lineNumber]);
    addReadingTempState.changeLines = addReadingTempState.changeLines.replace(regExpArr[lineNumber], '');
    const changeLinesIndex = addReadingTempState.changeLinesNumberArray.indexOf(`${lineNumber}`);
    if (changeLinesIndex !== -1) addReadingTempState.changeLinesNumberArray
      .splice(changeLinesIndex, 1);

    // console.log("cda:",this.state.changeLines);
    switch (coinsPoint) {
      case 6:
        side1 = classRedSide;
        side2 = classRedSide;
        middle1 = classMiddleB;
        middle2 = classRedMiddle;
        // erase the word if it has already exsited
        // console.log("changeLinesArr",this.changeLinesArr[lineNumber]);
        // this.state.changelines=this.state.changeLines.replace(this.changeLinesArr[lineNumber],"");
        addReadingTempState.changeLines += addReadingTempState.changeLines.length === 0 ? changeLinesArr[lineNumber] : `  ${changeLinesArr[lineNumber]}`;
        addReadingTempState.changeLinesNumberArray.push(lineNumber);// Also put the line number in an array
        // put number to image array
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = 7;
        break;
      case 7:
        side1 = classSide;
        middle1 = classMiddle;
        side2 = classSide;
        middle2 = classMiddle;
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = coinsPoint;
        break;
      case 9:
        side1 = classRedSide;
        side2 = classRedSide;
        middle1 = classRedMiddle;
        middle2 = classMiddleB;
        // erase the word if it has already exsited
        // this.state.changelines=this.state.changeLines.replace(this.changeLinesArr[lineNumber],"");
        addReadingTempState.changeLines += addReadingTempState.changeLines.length === 0 ? changeLinesArr[lineNumber] : `  ${changeLinesArr[lineNumber]}`;
        addReadingTempState.changeLinesNumberArray.push(lineNumber);// Also put the line number in an array
        // put number to image array
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = 8;
        break;
      case 8:
        side1 = classSide;
        side2 = classSide;
        middle1 = classMiddleB;
        middle2 = classMiddleB;
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = coinsPoint;
        break;
      default:
        break;
    }
    // console.log(`line${lineNumber}`);
    addReadingTempState[`line${lineNumber}`].side1 = side1;
    addReadingTempState[`line${lineNumber}`].side2 = side2;
    addReadingTempState[`line${lineNumber}`].middle1 = middle1;
    addReadingTempState[`line${lineNumber}`].middle2 = middle2;

    // set next line to be available
    addReadingTempState.availableArr[(lineNumber * 1) + 1] = true;
    // console.log(this.eliminateEmptyString(this.state.changeLines));
    // console.log(this.imageArrays);
    this.props.clickCoin(addReadingTempState); /* Make sure addReadingTempState is a new object */
  }

  /**
   * Handle subcomponent's submit callback.
   * @param {object} readingObject is the new reading object.
   * @returns {null} No return.
   */
  handleSubmitCallback(readingObject) {
    // if (event) event.preventDefault();
    // console.log("Submit");
    // locking screen and call api
    // this.setState({isWriting:true});
    // this.props.loading(true);
    const img1 = this.props.addReadingTempState.imageArrays.img1.join();
    const img2 = this.props.addReadingTempState.imageArrays.img2.join();

    const reading = {
      reading_name: readingObject.readingName,
      hexagram_arr_1: getImageArray(img1),
      hexagram_arr_2: getImageArray(img2),
      img1,
      img2,
      date: new Date(readingObject.date), // keeping a date object to mongoDB
      change_lines: this.props.addReadingTempState.changeLinesNumberArray,
      change_lines_text: AddReadingContainer
        .eliminateEmptyString(this.props.addReadingTempState.changeLines),
      people: readingObject.people,
      // user_id: LoginApi.isLogin(document).userid
    };
    // Before send to the server, setting up the right local time
    const currentTime = new Date();
    reading.date.setHours(currentTime.getHours());
    reading.date.setMinutes(currentTime.getMinutes());
    reading.date.setSeconds(currentTime.getSeconds());
    // Calling the action.
    this.props.createReading({ reading, jwtMessage: localStorage.getItem(JWT_MESSAGE) });
    /* Deprecated old version
    createReading(reading).then((result) => {
      // console.log("result:",result);
      this.props.loading(false);
      this.props.history.push('/reading');
    }); */
  }

  /**
   * The render method for this component.
   * @returns {null} No return.
   */
  render() {
    return (
      <div className="modal fade" id="addReadingModal" tabIndex="-1" role="dialog" aria-labelledby="addReadingModalLabel" aria-hidden="true">
        <div className="modal-dialog modal-lg" role="document">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title" id="addReadingModalLabel">Add a new reading</h5>
              <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                <span aria-hidden="true">&times;</span>
              </button>
            </div>
            <div className="modal-body">
              <LoadingAnimation />
              <AddReadingForm
                handleSubmit={(readingObject) => this.handleSubmitCallback(readingObject)}
                handleCancel={AddReadingContainer.handleCancelCallback}
                handleCoinClick={this.handleCoinClickCallback}
                addReadingTempState={this.props.addReadingTempState}
              />
            </div>
          </div>
        </div>
      </div>

    );
  }
}

const mapStateToProps = state => ({
  addReadingTempState: state.addReadingTempState,
  // isLoadingState: state.isLoading,
  user: state.user,
  readings: state.readings
});
const mapDispatchToProps = dispatch => ({
  clickCoin: addReadingTempState => dispatch(clickCoin(addReadingTempState)),
  clearAddReadingTempState: _ => dispatch(clearAddReadingTempState()),
  // loading: isLoadingBool => dispatch(isLoading(isLoadingBool)),
  checkAuthentication: _ => dispatch(checkAuthentication()),
  createReading: params => dispatch(createReading(params))
});
export default connect(mapStateToProps, mapDispatchToProps)(AddReadingContainer);
