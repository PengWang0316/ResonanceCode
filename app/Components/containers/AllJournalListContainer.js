import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import styles from '../../styles/AllJournalListContainer.module.css';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';
import JournalRow from '../JournalRow';
import Pagination from '../SharedComponents/Pagination';
import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import { checkAuthentication } from '../../actions/UserActions';
import { fetchAllJournal } from '../../actions/JournalActions';
import { NUMBER_PER_PAGE_JOURNAL } from '../../config';

/** Show all journal a user has. */
export class AllJournalListContainer extends Component {
  state = {
    journals: []
  };
  /** Checking the users' authentication state and calling fetchAllJournal function from JournalActions.
    * @return {null} No return.
  */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    else if (this.props.allJournal.length === 0) this.props.fetchAllJournal();
    // When the component is mounted second time, all props will remind the same. The componentWillReceivedProps method will not be called. (Not sure whether it is a bug.) So, the code below can add a checking to prevent this happen.
    if (this.state.journals.length === 0 && this.props.allJournal.length !== 0)
      this.setState({ journals: this.props.allJournal.slice(0, NUMBER_PER_PAGE_JOURNAL) });
    // this.currentPage = 0;
  }

  /** Calling fetchAllJournal function from JournalActions after the user login.
    * Setting the first showing journal to state.
    * @param {object} nextProps is an object that contains props' vaule.
    * @return {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if (!this.props.user.isAuth && nextProps.user.isAuth && this.props.allJournal.length === 0)
      this.props.fetchAllJournal();
    if (this.state.journals.length === 0 && nextProps.allJournal.length !== 0)
      this.setState({ journals: nextProps.allJournal.slice(0, NUMBER_PER_PAGE_JOURNAL) });
  }

  /** Getting the part of content from props.allJournal state and setting them to the component state.
    * @param {number} currentPage is the number of the page a user is clicking.
    * @return {null} No return.
  */
  changeContent = currentPage => {
    this.setState({
      journals: this.props.allJournal
        .slice(
          currentPage * NUMBER_PER_PAGE_JOURNAL,
          (currentPage * NUMBER_PER_PAGE_JOURNAL) + NUMBER_PER_PAGE_JOURNAL
        )
    });
  }

  /** Rendering jsx for the component.
    * @return {jsx} Return the jsx for the component.
  */
  render() {
    return (
      <UnauthenticatedUserCheck>
        <div className={`${styles.mainContentDiv}`}>
          <div className={`${styles.title}`}>Attached journal list</div>
          <div className="text-mute">(Not include <Link to={{ pathname: '/unattachedJournals', search: '?readingName=Unattached Journals' }}>unattached journals</Link>)</div>
          <LoadingAnimation />
          {/* Using the first reading id from the pingPongStates object. */}
          {this.state.journals.map(journal => (
            <JournalRow
              key={journal._id}
              journal={journal}
              readingId={Object.keys(journal.pingPongStates)[0]}
              isAllJournal
            />))}

          {this.props.allJournal.length !== 0 && <Pagination
            amount={this.props.allJournal.length}
            fetchContent={this.changeContent}
            numberPerpage={NUMBER_PER_PAGE_JOURNAL}
          />}
        </div>
      </UnauthenticatedUserCheck>
    );
  }
}

const mapStateToProps = state => ({
  user: state.user,
  allJournal: state.allJournal
});
const mapDispatchToProps = dispatch => ({
  fetchAllJournal: _ => dispatch(fetchAllJournal()),
  checkAuthentication: _ => dispatch(checkAuthentication())
});
export default connect(mapStateToProps, mapDispatchToProps)(AllJournalListContainer);
