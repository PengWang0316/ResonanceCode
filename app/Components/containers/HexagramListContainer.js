import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// import { checkAuthentication } from '../../actions/UserActions';
// import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import { TOTAL_NUMBER_HEXAGRAM } from '../../config';
import HexagramImage from '../HexagramImage';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';
import HexagramDetailModal from '../HexagramDetailModal';
import { fetchHexagrams, clearHexagrams } from '../../actions/HexagramActions';
import styles from '../../styles/HexagramListContainer.module.css';

/**
  Show a list for Hexagram
*/
export class HexagramListContainer extends Component {
  static propTypes = {
    hexagrams: PropTypes.array,
    fetchHexagrams: PropTypes.func.isRequired,
    clearHexagrams: PropTypes.func.isRequired
  };
  static defaultProps = { hexagrams: [] };
  state = { hexagram: null };
  /**
   * Fetching all hexagram is existed number of hexagrams less than 64.
   * @return {null} No return.
  */
  componentWillMount() {
    /* istanbul ignore next */
    if (this.props.hexagrams.length !== TOTAL_NUMBER_HEXAGRAM) {
      this.props.clearHexagrams();
      this.props.fetchHexagrams();
    }
  }

  /**
   * When a user clicks a row in the table, find the hexagram and show the modal.
   * @params {string} id repersents hexagram's id.
   * @return {null} No return.
   */
  handleHexagramClick = ({ target }) => {
    // Put all hexagram to a object and use id as the key.
    if (!this.hexagramsMap) {
      this.hexagramsMap = {};
      this.props.hexagrams.forEach(hexagram => { this.hexagramsMap[hexagram.number] = hexagram; });
    }
    // Because the event bubbling, we have to look up util reach the tr element.
    let id = null;
    let nextTarget = target.parentNode;
    while (!id) {
      if (nextTarget.nodeName === 'TR') ({ id } = nextTarget);
      else if (nextTarget.nodeName === 'TBODY') throw new Error('Missing id.');
      else nextTarget = nextTarget.parentNode;
    }
    this.setState({ hexagram: this.hexagramsMap[id] });
    $('#hexagramDetailModal').modal('toggle'); // $ will use jQuery from the index.html
  };

  /**
   * When the user click a associated hexagram, change the state.hexagram to that one.
   * @param {number} number is the hexagram number.
   * @return {null} No return.
   */
  handleAssociatedHexagramClick = number => this.setState({ hexagram: this.hexagramsMap[number] });

  /**
   * Rendering the jsx for the component.
   * @return {jsx} Returnt jsx.
   */
  render() {
    return (
      <div>
        <LoadingAnimation />
        <table>
          <thead className={styles.tableHead}>
            <tr>
              <th>Number</th>
              <th>Image</th>
              <th className={styles.rcName}><span>Resonance Code Name</span></th>
              <th className={styles.whhName}><span>Wilhelm/Huang/Hintley Name</span></th>
            </tr>
          </thead>
          <tbody>
            {this.props.hexagrams && this.props.hexagrams.map(hexagram => (
              <tr
                key={hexagram._id}
                id={hexagram.number}
                className={styles.tableRow}
                onClick={this.handleHexagramClick}
              >
                <td className="text-center">#{hexagram.number}</td>
                <td><HexagramImage imageNumber={hexagram.img_arr} isFirstImage isBlack /></td>
                <td className="pr-4">{hexagram.resonance_code_name}</td>
                <td>{hexagram.wilhelm_huang_hintley_name}</td>
              </tr>
            ))}
          </tbody>
        </table>
        <HexagramDetailModal
          hexagram={this.state.hexagram}
          handleHexagramClick={this.handleAssociatedHexagramClick}
        />
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  hexagrams: state.hexagrams
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchHexagrams: () => dispatch(fetchHexagrams({})),
  clearHexagrams: () => dispatch(clearHexagrams())
});
export default connect(mapStateToProps, mapDispatchToProps)(HexagramListContainer);
