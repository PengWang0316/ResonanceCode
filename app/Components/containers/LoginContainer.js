import React, { Component } from 'react';
import { connect } from 'react-redux';
import jquery from 'jquery';
import PropTypes from 'prop-types';

import styles from '../../styles/LoginContainer.module.css';
import LoginCoverImage from '../LoginCoverImage';
import LoginForm from '../LoginForm';
import { checkAuthentication } from '../../actions/UserActions';

/** The container for the login page. */
export class LoginContainer extends Component {
  static propTypes = {
    user: PropTypes.object.isRequired,
    checkAuthentication: PropTypes.func.isRequired
  };
  /** Checking the authentication when a user trys to load this page.
    * @return {null} No return.
  */
  componentWillMount() {
    // console.log(this.props.location.search);
    this.props.checkAuthentication(this.props.location.search); // Checking the authentication
  }

  /** Forwarding the page to reading after a user login successfully.
    * Showing the warning message when a user fails to login.
    * @param {object} nextProps is an object has new props that this component will receive.
    * @return {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    /* Pushing to the reading page after login */
    /* istanbul ignore next */
    if (nextProps.user.isAuth) this.props.history.push('/reading');
    else if (nextProps.user.loginErr) {
      jquery('#loginWarnMessage').css('opacity', '1');
      setTimeout(_ => { jquery('#loginWarnMessage').css('opacity', '0'); }, 3000);
    }
  }

  /** Render method for the component.
    * @return {jsx} Return the jsx for the component.
  */
  render() {
    return (
      <div className={`${styles.loginBackgroundContainer}`}>
        <LoginForm />
        <LoginCoverImage />
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({ user: state.user });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  checkAuthentication: jwt => dispatch(checkAuthentication(jwt))
});
export default connect(mapStateToProps, mapDispatchToProps)(LoginContainer);
