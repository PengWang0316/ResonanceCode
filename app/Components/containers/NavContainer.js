import React, { Component } from "react";
import Nav from "../Nav";

import {isLogin, logout} from "../../apis/LoginApi";

import { connect } from "react-redux";
// import { loginSuccess, logoutSuccess } from "../../actions/LoginActions";

class NavContainer extends Component {

  /*componentWillMount(){
    this.props.loginSuccess(isLogin(document));  send user imformation to state
  }
*/
  render(){
    return(
      <Nav user={isLogin(document)} handleLogout={()=>{logout(document);}} />
    );
  }

}

/*const mapStateToProps = (state, onwProps) => {  console.log("state:",state);
return {user: state.user};};
const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    loginSuccess: (user) => {dispatch(loginSuccess(user));},
    logoutSuccess: () => {dispatch(logoutSuccess());}
  }
};*/
export default NavContainer;
