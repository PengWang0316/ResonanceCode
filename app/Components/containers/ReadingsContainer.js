import React, { Component } from 'react';
import { connect } from 'react-redux';
// import QueryString from 'query-string';

import { NUMBER_OF_READING_PER_PAGE_RECENT_READINGS } from '../../config';
import { fetchRecentReadings, fetchReadingsAmount } from '../../actions/ReadingActions';
import { checkAuthentication } from '../../actions/UserActions';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';
import BriefReading from '../BriefReading';
import Pagination from '../SharedComponents/Pagination';
import AddReadingJournalButton from '../AddReadingJournalButton';
// import PageNavigationButton from '../PageNavigationButton';
import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import AddReadingContainer from './AddReadingContainer';
import DeleteReadingComformModal from '../DeleteReadingComformModal';
// import styles from '../styles/ReadingsContainer.module.css';

// import LoginApi from "../../apis/LoginApi";
// import DatabaseApi from "../../apis/DatabaseApi";

/**
 * The container for readings page.
 * @returns {null} return null.
 */
class ReadingsContainer extends Component {
  state = ({
    deleteReadingId: null,
    deleteReadingName: null
  });
  /**
 * Getting page info from url to decide showing which readings
 * and also check the authentication to decide whether fetch the readings.
 * @returns {null} return null.
 */
  componentWillMount() {
    // get the page number from url
    // this.setState({ isFinishedLoading: false });
    if (this.props.readingsAmount === null) this.props.fetchReadingsAmount();
    // const pageInfos = QueryString.parse(this.props.location.search);
    // if (pageInfos.start) this.startNumber = pageInfos.start;
    // this.startNumber = pageInfos.start ? pageInfos.start : '1';
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    else if (this.props.readings.length === 0) this.props.fetchRecentReadings(0);
  }
  /* componentWillMount() {
    // get the page number from url
    // this.setState({ isFinishedLoading: false });
    const pageInfos = QueryString.parse(this.props.location.search);
    if (pageInfos.start) this.startNumber = pageInfos.start;
    this.startNumber = pageInfos.start ? pageInfos.start : '1';
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    else if (this.props.readings.length === 0) this.props.fetchRecentReadings(this.startNumber);
  } */

  /**
   * Triger the fetch method after the user valids the authentication.
   * @param {object} user The object of user.
   * @returns {null} No return.
   */
  componentWillReceiveProps({ user, readings }) {
    if (!this.props.user.isAuth && user.isAuth && this.props.readings.length === 0) // Making sure the below code will be just loaded once.
      this.props.fetchRecentReadings(0);
    // else if (readings.length !== 0) this.setState({ isFinishedLoading: true });
  }

  /**
   * Showing the confrom modal to user.
   * @param {string} readingId is the id of reading that will be deleted.
   * @returns {null} No return.
   */
  handleDeleteCallback = ({ readingId, readingName }) => {
    // $ will use jQuery that comes from index.html
    this.setState({
      deleteReadingId: readingId,
      deleteReadingName: readingName
    });
    $('#deleteReadingConformModal').modal('toggle');
  }

  /**
   * Render method.
   * @returns {jsx} The jsx for ReadingsContainer page.
   */
  render() {
    return (
      <UnauthenticatedUserCheck>
        <div key="key_reading">
          <LoadingAnimation />
          {this.props.readings.map(reading => (
            <BriefReading
              key={reading._id}
              reading={reading}
              deleteReadingCallback={this.handleDeleteCallback}
            />))}

          {this.props.readings.length === 0 && !this.props.isLoading && <div className="font-weight-bold"><h4>There is no reading yet. Please add your reading.</h4></div>}
          {/* Old pagination component.
            <PageNavigationButton
            isEmptyContent={this.props.readings.length === 0}
            startNumber={this.startNumber}
          />
          */}
          {this.props.readingsAmount !== null && this.props.readingsAmount !== 0 && <Pagination
            amount={this.props.readingsAmount}
            fetchContent={this.props.fetchRecentReadings}
            numberPerpage={NUMBER_OF_READING_PER_PAGE_RECENT_READINGS}
          />}

          <AddReadingJournalButton />
        </div>

        <AddReadingContainer />
        <DeleteReadingComformModal
          readingId={this.state.deleteReadingId}
          readingName={this.state.deleteReadingName}
        />

      </UnauthenticatedUserCheck>
    );
  }
}
const mapStateToProps = (state, ownProps) => ({
  readings: state.readings,
  isLoading: state.isLoading,
  user: state.user,
  readingsAmount: state.readingsAmount
});
const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchRecentReadings: pageNumber => dispatch(fetchRecentReadings(pageNumber)),
  checkAuthentication: _ => dispatch(checkAuthentication()),
  fetchReadingsAmount: _ => dispatch(fetchReadingsAmount())
});
// const Readings = connect(mapStateToProps, mapDispatchToProps)(Readings);

export default connect(mapStateToProps, mapDispatchToProps)(ReadingsContainer);
