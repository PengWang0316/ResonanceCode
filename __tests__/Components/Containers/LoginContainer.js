import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { LoginContainer } from '../../../app/Components/containers/LoginContainer';

jest.mock('jquery', () => jest.fn());
jest.mock('../../../app/Components/LoginCoverImage', () => 'LoginCoverImage');
jest.mock('../../../app/Components/LoginForm', () => 'LoginForm');

describe('LoginContainer test', () => {
  const defaultProps = {
    user: { isAuth: true },
    checkAuthentication: jest.fn(),
    location: { search: 'location search' }
  };
  const getShallowComponent = (props = defaultProps) => shallow(<LoginContainer {...props} />);

  test('componentWillMount', () => {
    getShallowComponent();
    expect(defaultProps.checkAuthentication).toHaveBeenLastCalledWith('location search');
  });

  test('componentWillReceiveProps next user is authorized', () => {
    const mockPushFn = jest.fn();
    const component = getShallowComponent({ ...defaultProps, history: { push: mockPushFn } });
    component.setProps({ user: { isAuth: true } });
    expect(mockPushFn).toHaveBeenLastCalledWith('/reading');
  });

  test('componentWillReceiveProps has loginErr', () => {
    jest.useFakeTimers();
    const mockPushFn = jest.fn();
    const mockCssFn = jest.fn();
    const jquery = require('jquery');
    jquery.mockReturnValue({ css: mockCssFn });
    const component = getShallowComponent();
    component.setProps({ user: { loginErr: 'error message' } });
    expect(mockPushFn).not.toHaveBeenCalled();
    expect(jquery).toHaveBeenLastCalledWith('#loginWarnMessage');
    expect(jquery).toHaveBeenCalledTimes(1);
    expect(mockCssFn).toHaveBeenLastCalledWith('opacity', '1');
    expect(mockCssFn).not.toHaveBeenLastCalledWith('opacity', '0');
    jest.runAllTimers();
    expect(jquery).toHaveBeenCalledTimes(2);
    expect(mockCssFn).toHaveBeenLastCalledWith('opacity', '0');
  });

  test('LoginContainer snapshot', () => expect(renderer.create(<LoginContainer {...defaultProps} />).toJSON()).toMatchSnapshot());
});
