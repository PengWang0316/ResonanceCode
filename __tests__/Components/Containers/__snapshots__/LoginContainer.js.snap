// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainer test LoginContainer snapshot 1`] = `
<div
  className="loginBackgroundContainer"
>
  <LoginForm />
  <LoginCoverImage />
</div>
`;
